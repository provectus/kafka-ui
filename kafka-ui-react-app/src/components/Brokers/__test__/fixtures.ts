import { BrokersLogdirs } from 'generated-sources';

export const brokersPayload = [
  { id: 1, host: 'b-1.test.kafka.amazonaws.com', port: 9092 },
  { id: 2, host: 'b-2.test.kafka.amazonaws.com', port: 9092 },
];

export const clusterStatsPayload = {
  brokerCount: 2,
  activeControllers: 1,
  onlinePartitionCount: 138,
  offlinePartitionCount: 0,
  inSyncReplicasCount: 239,
  outOfSyncReplicasCount: 0,
  underReplicatedPartitionCount: 0,
  diskUsage: [
    { brokerId: 0, segmentSize: 334567, segmentCount: 245 },
    { brokerId: 1, segmentSize: 12345678, segmentCount: 121 },
  ],
  version: '2.2.1',
};

const partition = {
  broker: 2,
  offsetLag: 0,
  partition: 2,
  size: 0,
};
const topics = {
  name: '_confluent-ksql-devquery_CTAS_NUMBER_OF_TESTS_59-Aggregate-Aggregate-Materialize-changelog',
  partitions: [partition],
};

export const brokerLogDirsPayload: BrokersLogdirs[] = [
  {
    error: 'NONE',
    name: '/opt/kafka/data-0/logs',
    topics: [
      {
        ...topics,
        partitions: [partition, partition, partition],
      },
      topics,
      {
        ...topics,
        partitions: [],
      },
    ],
  },
];
