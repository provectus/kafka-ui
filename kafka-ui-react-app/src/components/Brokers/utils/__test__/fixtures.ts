import { BrokerMetrics } from 'generated-sources';

export const brokerMetricsPayload: BrokerMetrics = {
  segmentSize: 23,
  segmentCount: 23,
  metrics: [
    {
      name: 'TotalFetchRequestsPerSec',
      labels: {
        canonicalName:
          'kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics',
      },
      value: 10,
    },
    {
      name: 'ZooKeeperRequestLatencyMs',
      value: 11,
    },
    {
      name: 'RequestHandlerAvgIdlePercent',
    },
  ],
};
export const transformedBrokerMetricsPayload =
  '{"segmentSize":23,"segmentCount":23,"metrics":[{"name":"TotalFetchRequestsPerSec","labels":{"canonicalName":"kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics"},"value":10},{"name":"ZooKeeperRequestLatencyMs","value":11},{"name":"RequestHandlerAvgIdlePercent"}]}';
