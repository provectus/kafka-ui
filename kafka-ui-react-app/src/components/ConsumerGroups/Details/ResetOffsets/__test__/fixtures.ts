export const expectedOutputs = {
  EARLIEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'EARLIEST',
    topic: '__amazon_msk_canary',
  },
  LATEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'LATEST',
    topic: '__amazon_msk_canary',
  },
  OFFSET: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: '10',
        partition: 0,
      },
    ],
    resetType: 'OFFSET',
    topic: '__amazon_msk_canary',
  },
};
