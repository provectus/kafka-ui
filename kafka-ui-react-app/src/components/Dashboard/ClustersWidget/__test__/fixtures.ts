import { Cluster, ServerStatus } from 'generated-sources';

export const onlineCluster: Cluster = {
  name: 'secondLocal',
  defaultCluster: false,
  status: ServerStatus.Online,
  brokerCount: 1,
  onlinePartitionCount: 6,
  topicCount: 3,
  bytesInPerSec: 0.000030618196853764715,
  bytesOutPerSec: 5.737800890036267075817,
};

export const offlineCluster: Cluster = {
  name: 'local',
  defaultCluster: true,
  status: ServerStatus.Offline,
  brokerCount: 1,
  onlinePartitionCount: 2,
  topicCount: 2,
  bytesInPerSec: 8000.0000067376808542600021,
  bytesOutPerSec: 0.8153063567297119490871,
};

export const clusters: Cluster[] = [onlineCluster, offlineCluster];
