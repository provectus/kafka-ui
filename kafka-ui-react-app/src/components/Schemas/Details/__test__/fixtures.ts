import { SchemaSubject, SchemaType } from 'generated-sources';
import {
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
} from 'redux/reducers/schemas/__test__/fixtures';

export const versionPayload = [
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
];
export const versionEmptyPayload = [];

export const jsonSchema: SchemaSubject = {
  subject: 'test',
  version: '15',
  id: 1,
  schema:
    '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.JSON,
};

export const protoSchema: SchemaSubject = {
  subject: 'test_proto',
  version: '1',
  id: 2,
  schema:
    'syntax = "proto3";\npackage com.indeed;\n\nmessage MyRecord {\n  int32 id = 1;\n  string name = 2;\n}\n',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.PROTOBUF,
};
