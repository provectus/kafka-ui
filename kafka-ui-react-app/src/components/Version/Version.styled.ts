import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
  display: flex;
  align-items: baseline;
`;

const textStyle = css`
  font-family: Inter, sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 20px;
`;

export const CurrentVersion = styled.span(
  ({ theme }) => css`
    ${textStyle};
    color: ${theme.version.currentVersion.color};
    margin-left: 0.25rem;
  `
);

export const OutdatedWarning = styled.span`
  ${textStyle}
`;

export const CurrentCommitLink = styled.a(
  ({ theme }) => css`
    ${textStyle};
    color: ${theme.version.commitLink.color};
    margin-left: 0.25rem;
    &:hover {
      color: ${theme.version.commitLink.color};
    }
  `
);
