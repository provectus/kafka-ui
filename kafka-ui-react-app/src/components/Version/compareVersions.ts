const split = (v: string): string[] => {
  const c = v.replace('v', '').split('-')[0];
  return c.split('.');
};

const compareVersions = (v1?: string, v2?: string): number => {
  if (!v1 || !v2) return 0;

  // try..catch - is our safeguard for strange git tags (or usecases without network)
  try {
    const s1 = split(v1);
    const s2 = split(v2);

    for (let i = 0; i < Math.max(s1.length, s2.length); i += 1) {
      const n1 = parseInt(s1[i] || '0', 10);
      const n2 = parseInt(s2[i] || '0', 10);

      if (n1 > n2) return 1;
      if (n2 > n1) return -1;
    }

    return 0;
  } catch (_) {
    return 0;
  }
};

export default compareVersions;
