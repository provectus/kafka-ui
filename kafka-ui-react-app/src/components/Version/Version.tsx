import React from 'react';
import WarningIcon from 'components/common/Icons/WarningIcon';
import { gitCommitPath } from 'lib/paths';
import { useActuatorInfo } from 'lib/hooks/api/actuatorInfo';
import { BUILD_VERSION_PATTERN } from 'lib/constants';
import { useLatestVersion } from 'lib/hooks/api/latestVersion';
import { formatTimestamp } from 'lib/dateTimeHelpers';

import * as S from './Version.styled';
import compareVersions from './compareVersions';

const Version: React.FC = () => {
  const { data: actuatorInfo = {} } = useActuatorInfo();
  const { data: latestVersionInfo = {} } = useLatestVersion();

  const tag = actuatorInfo?.build?.version;
  const commit = actuatorInfo?.git?.commit.id;
  const { tag_name: latestTag } = latestVersionInfo;

  const outdated = compareVersions(tag, latestTag);

  const currentVersion = tag?.match(BUILD_VERSION_PATTERN)
    ? tag
    : formatTimestamp(actuatorInfo?.build?.time);

  if (!tag) return null;

  return (
    <S.Wrapper>
      {!!outdated && (
        <S.OutdatedWarning
          title={`Your app version is outdated. Current latest version is ${latestTag}`}
        >
          <WarningIcon />
        </S.OutdatedWarning>
      )}

      {commit && (
        <div>
          <S.CurrentCommitLink
            title="Current commit"
            target="__blank"
            href={gitCommitPath(commit)}
          >
            {commit}
          </S.CurrentCommitLink>
        </div>
      )}
      <S.CurrentVersion>{currentVersion}</S.CurrentVersion>
    </S.Wrapper>
  );
};

export default Version;
