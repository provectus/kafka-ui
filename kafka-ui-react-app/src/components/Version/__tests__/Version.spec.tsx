import React from 'react';
import { screen } from '@testing-library/dom';
import Version from 'components/Version/Version';
import { render } from 'lib/testHelpers';
import { formatTimestamp } from 'lib/dateTimeHelpers';
import { useActuatorInfo } from 'lib/hooks/api/actuatorInfo';
import { useLatestVersion } from 'lib/hooks/api/latestVersion';
import { actuatorInfoPayload } from 'lib/fixtures/actuatorInfo';
import { latestVersionPayload } from 'lib/fixtures/latestVersion';

jest.mock('lib/hooks/api/actuatorInfo', () => ({
  useActuatorInfo: jest.fn(),
}));
jest.mock('lib/hooks/api/latestVersion', () => ({
  useLatestVersion: jest.fn(),
}));

describe('Version Component', () => {
  const versionTag = 'v0.5.0';
  const snapshotTag = 'test-SNAPSHOT';
  const commitTag = 'befd3b328e2c9c7df57b0c5746561b2f7fee8813';

  const actuatorVersionPayload = actuatorInfoPayload(versionTag);
  const formattedTimestamp = formatTimestamp(actuatorVersionPayload.build.time);

  beforeEach(() => {
    (useActuatorInfo as jest.Mock).mockImplementation(() => ({
      data: actuatorVersionPayload,
    }));
    (useLatestVersion as jest.Mock).mockImplementation(() => ({
      data: latestVersionPayload,
    }));
  });

  describe('tag does not exist', () => {
    it('does not render component', async () => {
      (useActuatorInfo as jest.Mock).mockImplementation(() => ({
        data: null,
      }));
      const { container } = render(<Version />);
      expect(container.firstChild).toBeEmptyDOMElement();
    });
  });

  describe('renders current version', () => {
    it('renders release build version as current version', async () => {
      render(<Version />);
      expect(screen.getByText(versionTag)).toBeInTheDocument();
    });
    it('renders formatted timestamp as current version when version is commit', async () => {
      (useActuatorInfo as jest.Mock).mockImplementation(() => ({
        data: actuatorInfoPayload(commitTag),
      }));
      render(<Version />);
      expect(screen.getByText(formattedTimestamp)).toBeInTheDocument();
    });
    it('renders formatted timestamp as current version when version contains -SNAPSHOT', async () => {
      (useActuatorInfo as jest.Mock).mockImplementation(() => ({
        data: actuatorInfoPayload(snapshotTag),
      }));
      render(<Version />);
      expect(screen.getByText(formattedTimestamp)).toBeInTheDocument();
    });
  });

  describe('outdated build version', () => {
    it('renders warning message', async () => {
      (useActuatorInfo as jest.Mock).mockImplementation(() => ({
        data: actuatorInfoPayload('v0.3.0'),
      }));
      render(<Version />);
      expect(
        screen.getByTitle(
          `Your app version is outdated. Current latest version is ${latestVersionPayload.tag_name}`
        )
      ).toBeInTheDocument();
    });
  });

  describe('current commit id with link', () => {
    it('renders', async () => {
      render(<Version />);
      expect(
        screen.getByText(actuatorVersionPayload.git.commit.id)
      ).toBeInTheDocument();
    });
  });
});
