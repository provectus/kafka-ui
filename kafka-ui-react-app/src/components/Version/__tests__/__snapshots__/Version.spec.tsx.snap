// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Version matches snapshot 1`] = `
<Version
  commit="123sdf34"
  tag="v1.0.1-SHAPSHOT"
>
  <div
    className="is-size-8 has-text-grey"
  >
    <span
      className="has-text-grey-light mr-1"
    >
      Version:
    </span>
    <span
      className="mr-1"
    >
      v1.0.1-SHAPSHOT
    </span>
    <span>
      (
    </span>
    <a
      href="https://github.com/provectus/kafka-ui/commit/123sdf34"
      target="__blank"
      title="Current commit"
    >
      123sdf34
    </a>
    <span>
      )
    </span>
  </div>
</Version>
`;
